### PR TITLE
feat: added pagination and filter for addresses and contacts

### PR DESCRIPTION
**Pull request by MyuddinKhatri**
- [x]  added pagination and filter for addresses and contacts
- [x] moved address and contact link and delete customizations to frappe 
screencast-jhaudio_8008-2021 07 04-20_28_13

https://github.com/Bloomstack/erpnext/pull/1403
https://github.com/Bloomstack/bloomstack_core/pull/611

### CODE REVIEW

**Comment by MyuddinKhatri** [comment]
@hrwX , @DeveloperMetal , @nehasacher143 I need to move the search Contact and Search address field above the listing, can you guys suggest me something?

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 1]
```diff
@@ -1,47 +1,312 @@
...
 $.extend(frappe.contacts, {
-	clear_address_and_contact: function(frm) {
-		$(frm.fields_dict['address_html'].wrapper).html("");
+	clear_address_and_contact: function (frm) {
+		frm.fields_dict['address_html'] && $(frm.fields_dict['address_html'].wrapper).html("");
 		frm.fields_dict['contact_html'] && $(frm.fields_dict['contact_html'].wrapper).html("");
 	},
 
-	render_address_and_contact: function(frm) {
+	paginate_address: function (frm) {
```
Comment (hrwX) [inline frappe/public/js/frappe/utils/address_and_contact.js]: - Commonify paginate_contact and paginate_address

**Review by hrwX — CHANGES_REQUESTED** [review]

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 2]
```diff
@@ -1,47 +1,312 @@
...
+		$('div[title="contact_html"]').append("<div class='contact_pagination'><p>Hello</p></div>");
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Do you realize that this selector: `$('div[title="contact_html"]')`
Will match any div with title "contact_html" even if the dom is hidden or behind the current page?

It means that you are potentially touching dom outside of the current page you expect this to work on. Please scope your selectors to the current container.

Not sure about the frappe.contacts instance but most classes have either a $page or $container you can use to scope things and not bleed behavior onto other similar selectors.

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 3]
```diff
@@ -1,47 +1,312 @@
...
+		}
+		for(var i=1; i<=count; i++){
+		  $(".address_pagination").append("<div class='page_number'>"+i+"</div>");
+		}
+		$(".page_number:first-child").addClass('active');
+		$("div[data-fieldname='address_html'] .address-box").filter(function(index) {
+			return index < address_per_page;
+		}).addClass('active');
+
+		$('body').on('click',`.page_number`,function(){
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Same issue here on container scope and performance. Don't attach live event onto the body. It means that jquery will scan the entire dom every time an element is added for that **.page_number** class.

Scope it to the page or even the field container, must faster to iterate.

**Review by DeveloperMetal — CHANGES_REQUESTED** [review]
Move the search box above the addresses component. You can prepend a container update to do your work.

And as noted scope all your selectors, fetching general classes kills page performance and will add odd bugs on other parts of the page which are hidden and later displayed.

**Comment by sahil28297** [comment]
@MyuddinKhatri What's up with this PR? Also fix Sider issues

**Comment by MyuddinKhatri** [comment]
> @MyuddinKhatri What's up with this PR? Also fix Sider issues

need to change some css here

**Comment by MyuddinKhatri** [comment]
@hrwX and @sahil28297 changes are done please have a look.

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 4]
```diff
@@ -1,47 +1,298 @@
 frappe.provide('frappe.contacts')
 
 $.extend(frappe.contacts, {
-	clear_address_and_contact: function(frm) {
-		$(frm.fields_dict['address_html'].wrapper).html("");
+	clear_address_and_contact: function (frm) {
+		frm.fields_dict['address_html'] && $(frm.fields_dict['address_html'].wrapper).html("");
 		frm.fields_dict['contact_html'] && $(frm.fields_dict['contact_html'].wrapper).html("");
+		frm.fields_dict['address_html1'] && $(frm.fields_dict['address_html1'].wrapper).html("");
+		frm.fields_dict['contact_html1'] && $(frm.fields_dict['contact_html1'].wrapper).html("");
```
Comment (hrwX) [inline frappe/public/js/frappe/utils/address_and_contact.js]: - `address_html1` is not available in frm.

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 5]
```diff
@@ -1,47 +1,298 @@
...
 	},
 
-	render_address_and_contact: function(frm) {
+
+
+	paginate: function (total_items, pagination_class, html_class, box_class) {
+		let per_page = 3;
```
Comment (hrwX) [inline frappe/public/js/frappe/utils/address_and_contact.js]:
```suggestion
		let per_page = 5;
```

**Review by hrwX — CHANGES_REQUESTED** [review]

**Review by hrwX — DISMISSED** [review]

**Comment by hrwX** [comment]
@MyuddinKhatri Fix sider issues.

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 6]
```diff
@@ -1,47 +1,274 @@
...
-		if (last_route && last_route.length > 3)
-			docname = last_route.slice(2).join("/");
-
-		return {
-			doctype,
-			docname
-		}
+		});
+
+		$(document).on('click', '.contact-box a.delete_contact', function () {
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Scope this please, you are forcing jquery to traverse the entire dom to set this event

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 7]
```diff
@@ -1,47 +1,274 @@
...
+						"link_name": frm.docname
+					},
+					callback: function () {
+						window.location.reload();
+					}
+				});
+			}, __("Select Contact"));
+		});
+
+		$(document).on('click', '.contact-box a.unlink_contact', function () {
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Scope this please, you are forcing jquery to traverse the entire dom to set this event

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 8]
```diff
@@ -1,47 +1,274 @@
...
+				};
+				$(frm.layout_for_contacts.fields_dict["contact_html1"].wrapper).html(frappe.render_template("contact_list", contact_list));
+				let no_of_contacts = contact_list.contact_list.length;
+				frappe.contacts.paginate(no_of_contacts, "contact_pagination", "contact_html1", "contact-box");
+			});
+		}
+
+		let no_of_contacts = frm.doc.__onload.contact_list.length;
+		frappe.contacts.paginate(no_of_contacts, "contact_pagination", "contact_html1", "contact-box");
+		$(document).on('click', '.btn-contact-link', function () {
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Scope this please, you are forcing jquery to traverse the entire dom to set this event

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 9]
```diff
@@ -1,47 +1,274 @@
...
+						},
+						callback: function () {
+							window.location.reload();
+						}
+					});
+				}
+			);
+		});
+
+		$(document).on('click', '.address-box a.delete_address', function () {
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Scope this please, you are forcing jquery to traverse the entire dom to set this event

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 10]
```diff
@@ -1,47 +1,274 @@
...
+						"link_name": frm.docname
+					},
+					callback: function () {
+						window.location.reload();
+					}
+				});
+			}, __("Select Address"));
+		});
+
+		$(document).on('click', '.address-box a.unlink_address', function () {
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Scope this please, you are forcing jquery to traverse the entire dom to set this event

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 11]
```diff
@@ -1,47 +1,274 @@
...
+				};
+				$(frm.layout_for_address.fields_dict["address_html1"].wrapper).html(frappe.render_template("address_list", addr_list));
+				let no_of_addresses = addr_list.addr_list.length;
+				frappe.contacts.paginate(no_of_addresses, "address_pagination", "address_html1", "address-box");
+			});
 		}
 
+		let no_of_addresses = frm.doc.__onload.addr_list.length;
+		frappe.contacts.paginate(no_of_addresses, "address_pagination", "address_html1", "address-box");
+		$(document).on('click', '.btn-address-link', function () {
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Scope this please, you are forcing jquery to traverse the entire dom to set this event

File: frappe/public/js/frappe/utils/address_and_contact.js [hunk 12]
```diff
@@ -1,47 +1,274 @@
...
+		}
+		for (var i = 1; i <= count; i++) {
+			$(`.${pagination_class}`).append("<div class='page_number'>" + i + "</div>");
+		}
+		$(`.${pagination_class} .page_number:first-child`).addClass('active');
+		$(`div[data-fieldname=${html_class}] .${box_class}`).filter(function (index) {
+			return index < per_page;
+		}).addClass('active');
+
+		$(`body`).on('click', `.${pagination_class} .page_number`, function () {
```
Comment (DeveloperMetal) [inline frappe/public/js/frappe/utils/address_and_contact.js]: Scope this please, you are forcing jquery to traverse the entire dom to set this event

**Review by DeveloperMetal — CHANGES_REQUESTED** [review]
I still see jquery selector scoping issues that can hurt browser performance. The .on(<event>, <class>, <callback>) method will trigger every time an element is added to the dom specially from the document scope. You can limit that traversal by scoping to the current page.

**Review by nehasacher143 — DISMISSED** [review]
@MyuddinKhatri  Looks good to me

@DeveloperMetal @hrwX Please review

File: frappe/contacts/address_and_contact.py [hunk 13]
```diff
@@ -178,3 +179,84 @@ def set_link_title(doc):
 		if not link.link_title:
 			linked_doc = frappe.get_doc(link.link_doctype, link.link_name)
 			link.link_title = linked_doc.get("title_field") or linked_doc.get("name")
+
+@frappe.whitelist()
+def get_addresses(doc, search_keyword):
+	doc = json.loads(doc)
```
Comment (hrwX) [inline frappe/contacts/address_and_contact.py]:
```suggestion
	doc = frappe.parse_json(doc)
```

File: frappe/contacts/address_and_contact.py [hunk 14]
```diff
@@ -178,3 +179,84 @@ def set_link_title(doc):
...
+
+	from frappe.contacts.doctype.address.address import get_address_display
+	address_list = [a.update({"display": get_address_display(a)})
+		for a in address_list]
+	return address_list
+
+@frappe.whitelist()
+def get_contacts(doc, search_keyword):
+	from frappe.contacts.doctype.address.address import get_condensed_address
+	doc = json.loads(doc)
```
Comment (hrwX) [inline frappe/contacts/address_and_contact.py]:
```suggestion
	doc = frappe.parse_json(doc)
```

**Review by hrwX — CHANGES_REQUESTED** [review]

**Review by hrwX — DISMISSED** [review]

**Review by DeveloperMetal — DISMISSED** [review]

File: frappe/contacts/address_and_contact.py [hunk 15]
```diff
@@ -7,6 +7,7 @@
 from frappe import _
 import functools
 import re
+import json
```
Comment (sahil28297) [inline frappe/contacts/address_and_contact.py]:
```suggestion
```

**Review by sahil28297 — CHANGES_REQUESTED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**